src/routes/_surfaces/workspace/explore/time-controls/TimeGrainSelector.svelte

```diff
@@ -0,0 +1,105 @@
+<script lang="ts">
+  import type {
+    TimeGrain,
+    TimeRangeName,
+  } from "$common/database-service/DatabaseTimeSeriesActions";
+  import CaretDownIcon from "$lib/components/icons/CaretDownIcon.svelte";
+  import Menu from "$lib/components/menu/Menu.svelte";
+  import MenuItem from "$lib/components/menu/MenuItem.svelte";
+  import FloatingElement from "$lib/components/tooltip/FloatingElement.svelte";
+  import { getMetricsExplorerById } from "$lib/redux-store/explore/explore-readables";
+  import { onClickOutside } from "$lib/util/on-click-outside";
+  import { tick } from "svelte";
+  import {
+    getDefaultTimeGrain,
+    getSelectableTimeGrains,
+    prettyTimeGrain,
+  } from "./time-range-utils";
+
+  export let metricsDefId: string;
+  export let selectedTimeRangeName: TimeRangeName;
+  export let selectedTimeGrain: TimeGrain;
+  export let onSelectTimeGrain: (timeGrain: TimeGrain) => void;
+
+  $: metricsExplorer = getMetricsExplorerById(metricsDefId);
+
+  let selectableTimeGrains: TimeGrain[];
+
+  // TODO: replace this with a call to the `/meta` endpoint, once available.
+  $: if (selectedTimeRangeName && $metricsExplorer?.allTimeRange) {
+    selectableTimeGrains = getSelectableTimeGrains(
+      selectedTimeRangeName,
+      $metricsExplorer.allTimeRange
+    );
+  }
+
+  // Set the default time grain when:
+  // - there's no time grain currently selected
+  // - the selected time grain is not in the list of selectable time grains (which can happen when the time range name is changed)
+  $: if (
+    !selectedTimeGrain ||
+    selectableTimeGrains.find(
+      (timeGrain) => timeGrain === selectedTimeGrain
+    ) === undefined
+  ) {
+    selectedTimeGrain = getDefaultTimeGrain(selectedTimeRangeName);
+    onSelectTimeGrain(selectedTimeGrain);
```

Comment:
We should dispatch instead of passing callbacks.

Reply:
Yeah, sounds good. Hamilton made the same comment 👍